app/services/integration/kubernetes_service.rb

```diff
@@ -12,6 +12,10 @@ def initialize(integration, namespace: self.class.namespace)
     @client = K8s::Client.autoconfig(namespace: namespace).extend(MergePatch)
   end
 
+  def use_openshift_route?
+    return client.api_groups.include?('route.openshift.io/v1') && !ActiveModel::Type::Boolean.new.cast(ENV['FORCE_NATIVE_INGRESS'])
+  end
```

Comment:
This is the line I am referring to: https://github.com/3scale/zync/blob/master/app/services/discover_integration_service.rb#L34

It seems that after this is frozen I can no longer update `@use_openshift_route`.

Reply:
As I said the existing tests didn't freeze the `Integration::KubernetesService` so the error didn't occur when the test ran. The error only occurred when the job ran because of the line referenced above. And the job error didn't print the stack trace. So to reproduce this in test I changed the test so it also froze the object as well.

In the test:
```
@service = Integration::KubernetesService.new(nil).freeze
```

Then with the snippet you provided:
```
def use_openshift_route?
    @use_openshift_route ||= !Rails.application.config.integrations.fetch("kubernetes_force_native_ingress", false) && client.api_groups.include?('route.openshift.io/v1')
end
```

Output from test (it fails 3 tests with the same error):
```
1) Failure:
Integration::KubernetesServiceTest#test_route_status_missing_ingress [/opt/app-root/test/services/kubernetes_service_test.rb:248]:
[Integration::KubernetesService::MissingStatusIngress] exception expected, not
Class: <FrozenError>
Message: <"can't modify frozen Integration::KubernetesService: #<Integration::KubernetesService:0x00005611ce2c0b88 @integration=nil, @namespace=\"zync\", @client=#<K8s::Client:0x00005611ce2ba198 @transport=#<K8s::Transport:0x00005611ce2ba5a8 @server=\"http://localhost:80\", @path_prefix=\"/\", @auth_token=\"token\", @auth_username=nil, @auth_password=nil, @options={:ssl_cert_store=>#<OpenSSL::X509::Store:0x00005611ce2bacb0 @verify_callback=nil, @error=nil, @error_string=nil, @chain=nil, @time=nil, @_httpclient_cert_store_items=[:default, #<OpenSSL::X509::Certificate: subject=#<OpenSSL::X509::Name CN=root-ca>, issuer=#<OpenSSL::X509::Name CN=root-ca>, serial=#<OpenSSL::BN:0x00005611ce1483a0>, not_before=2019-04-04 11:32:59 UTC, not_after=2029-04-01 11:32:59 UTC>]>}, @logger=#<Logger:0x00005611ce2ba2d8 @level=0, @progname=\"K8s::Transport<http://localhost:80>\", @default_formatter=#<Logger::Formatter:0x00005611ce2ba2b0 @datetime_format=nil>, @formatter=nil, @logdev=#<Logger::LogDevice:0x00005611ce2ba260 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDERR>>, @binmode=false, @mon_data=#<Monitor:0x00005611ce2ba238>, @mon_data_owner_object_id=19200>>, @excon=#<Excon::Connection:9650 @data={:chunk_size=>1048576, :ciphers=>\"ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA256:DHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:ECDHE-ECDSA-DES-CBC3-SHA:ECDHE-RSA-DES-CBC3-SHA:EDH-RSA-DES-CBC3-SHA:DES-CBC3-SHA:!DSS\", :connect_timeout=>60, :debug_request=>false, :debug_response=>false, :headers=>{\"Accept\"=>\"application/json\"}, :idempotent=>false, :instrumentor_name=>\"excon\", :middlewares=>[Excon::Middleware::RedirectFollower, Excon::Middleware::ResponseParser, Excon::Middleware::Expects, Excon::Middleware::Idempotent, Excon::Middleware::Instrumentor, Excon::Middleware::Mock], :mock=>true, :nonblock=>true, :omit_default_port=>false, :persistent=>true, :read_timeout=>60, :retry_errors=>[Excon::Error::Timeout, Excon::Error::Socket, Excon::Error::HTTPStatus], :retry_limit=>4, 
:ssl_verify_peer=>true, :ssl_uri_schemes=>[\"https\"], :stubs=>:global, :tcp_nodelay=>false, :thread_safe_sockets=>true, :uri_parser=>URI, :versions=>\"excon/0.71.1 (x86_64-linux) ruby/2.7.4\", :write_timeout=>60, :host=>\"localhost\", :hostname=>\"localhost\", :path=>\"\", :port=>80, :query=>nil, :scheme=>\"http\", :ssl_cert_store=>#<OpenSSL::X509::Store:0x00005611ce2bacb0 @verify_callback=nil, @error=nil, @error_string=nil, @chain=nil, @time=nil, @_httpclient_cert_store_items=[:default, #<OpenSSL::X509::Certificate: subject=#<OpenSSL::X509::Name CN=root-ca>, issuer=#<OpenSSL::X509::Name CN=root-ca>, serial=#<OpenSSL::BN:0x00005611ce13f070>, not_before=2019-04-04 11:32:59 UTC, not_after=2029-04-01 11:32:59 UTC>]>, :__construction_args=>{:host=>\"localhost\", :hostname=>\"localhost\", :path=>\"\", :port=>80, :query=>nil, :scheme=>\"http\", :persistent=>true, :middlewares=>[Excon::Middleware::RedirectFollower, Excon::Middleware::ResponseParser, Excon::Middleware::Expects, Excon::Middleware::Idempotent, Excon::Middleware::Instrumentor, Excon::Middleware::Mock], :headers=>{\"Accept\"=>\"application/json\"}, :ssl_cert_store=>#<OpenSSL::X509::Store:0x00005611ce2bacb0 @verify_callback=nil, @error=nil, @error_string=nil, @chain=nil, @time=nil, @_httpclient_cert_store_items=[:default, #<OpenSSL::X509::Certificate: subject=#<OpenSSL::X509::Name CN=root-ca>, issuer=#<OpenSSL::X509::Name 
CN=root-ca>, serial=#<OpenSSL::BN:0x00005611ce13e800>, not_before=2019-04-04 11:32:59 UTC, not_after=2029-04-01 11:32:59 UTC>]>}} @socket_key=\"http://localhost:80\" @_excon_sockets={9480=>{}} @persistent_socket_reusable=true>>, @namespace=\"zync\", @api_clients={}, @mon_data=#<Monitor:0x00005611ce2ba0d0>, @mon_data_owner_object_id=19220, @api_groups=[\"networking.k8s.io/v1\", \"route.openshift.io/v1\"]>>">
---Backtrace---
app/services/integration/kubernetes_service.rb:18:in `use_openshift_route?'
app/services/integration/kubernetes_service.rb:396:in `persist_provider'
app/services/integration/kubernetes_service.rb:46:in `handle_provider'
app/services/integration/kubernetes_service.rb:37:in `call'
test/services/kubernetes_service_test.rb:249:in `block (2 levels) in <class:KubernetesServiceTest>'
test/services/kubernetes_service_test.rb:248:in `block in <class:KubernetesServiceTest>'
---------------
```

This is the same error from the job error but with an included stack trace. It seems clear to me that the issue is that the object is frozen but this method is trying to update the property.

Reply:
Yes. Sorry, didn't understand what you meant in previous comment. I don't see an important reason to have the object frozen.

This can be fixed I guess by stop freezing the whole object, or by using an intermediary mutable object to store the value.

Reply:
Or because kubernetes client is using same server always (only different namespaces), perhaps a class instance variable can do the job.

```ruby

class << self
  attr_accessor :native_ingress

  def native_ingress_discovered?
    defined? :@native_ingress
  end
end

def use_openshift_route?
  return self.native_ingress if self.native_ingress_discovered?

  self.native_ingress = !Rails.application.config.integrations.fetch("kubernetes_force_native_ingress", false) && client.api_groups.include?('route.openshift.io/v1')
end

Note that what I originally suggested to use `||=` operator is not good as it will repeat trying in case value is `false`, not only for `nil`.

Still I don't see a problem not to freeze the object btw. But don't have much time now for this issue and investigating what would work best.

Reply:
Thanks @akostadinov, I think I understand the problem and I understand you have other priorities. 

I wasn't able to implement your suggestion above, because `self.native_ingress_discovered?` threw an undefined method error. I also found it a bit difficult to reason about. Where does this sub class `self` get initialized? Is it a sub class? I am not familiar with how the nested class syntax works in Ruby. I suspect if I understood this I would be able to resolve this error, so please forgive my ignorance here. 

In any case, I came up with a slightly different solution that did work and made a bit more sense to me. How do you feel about this?

```ruby
  def initialize(integration, namespace: self.class.namespace)
    super(integration)
    @namespace = namespace
    @client = K8s::Client.autoconfig(namespace: namespace).extend(MergePatch).extend(UseOpenshiftRoute)
  end

  module UseOpenshiftRoute
    attr_accessor :use_openshift_route
    
    def use_openshift_route?
      return use_openshift_route unless use_openshift_route.nil?
      @use_openshift_route = !Rails.application.config.integrations.fetch(:kubernetes_force_native_ingress, false) && api_groups.include?('route.openshift.io/v1')
    end
  end

  ...
  if client.use_openshift_route?
    # do x
  else
    # do y
  end
  ...
```

It's very similar to your suggestion but I used the `nil? method to avoid the need for an additional variable and then I placed the method on the K8S client instead following the same convention as the `MergePatch` extension. Since this object is not deeply frozen this looks to work just fine. And it make sense to me that this information is present on the client.

I didn't really want to change the other code to not freeze the object because I really don't know what effect that will have outside of my code changes. I agree that it's probably not a problem but I would hate to cause something unexpected. In fact, I will likely leave the freeze in the test to avoid anyone else making a change here and running into the same issue I did at runtime.

Reply:
Ok I made a few changes because setting the property on the name `native_ingress` didn't make sense because it's actually the opposite. Also the extra property was not needed. 

```ruby
  class << self
    attr_accessor :use_openshift_route
  end
  
  def use_openshift_route?
    return self.class.use_openshift_route unless self.class.use_openshift_route.nil?
  
    self.class.use_openshift_route = !Rails.application.config.integrations.fetch(:kubernetes_force_native_ingress, false) && client.api_groups.include?('route.openshift.io/v1')
  end
```

At first I couldn't figure out how to clear the setting between tests (i.e. openshift/native ingress) so I had to do some more research to understand this syntax. After researching how this `class << self` works, I found out that this is a singleton for the class itself, not the object. In other languages this would be like a static method/property. Once I understood this, I was able to get the tests to work. 

If the above looks ok to you, I will commit my changes to this PR.